### PR TITLE
allow default shell rc path to be extended via FLUX_SHELL_RC_PATH

### PR DESCRIPTION
**Pull request by grondo**
This is an attempt at a simple solution for #3866.

The default `initrc.lua` for the job shell loads all `*.lua` rc scripts from `${rcpath}/lua.d/*`, where `rcpath` is the directory in which `initrc.lua` was found. On a normal system, the default shell initialization can be influenced by dropping another Lua rc script in this directory. (This is how the mpi 'personality' plugins work).

However, in systems like spack, where every package has its own root, add-on packages like mpibind can't just drop a new file into the shell `rcpath`, since that path is in the flux-core package's root. This PR adds support for a new environment variable `FLUX_SHELL_RC_PATH` which, when set in a job's environment, will extend the search path for `*.lua` files to automatically load at runtime.

Therefore, in spack, packages can append a path to a directory where they keep a Lua rc file which can then load plugins, further change environment, etc.

For example, the mpibind spack package, which wants to always load the `mpibind.so` shell plugin, can install a one-liner `mpibind.lua` like:
```Lua
plugin.load ("/full/path/to/mpibind.so")
```
and in the environment manipulation for the spack package, extend the environment to append the directory in which this Lua file is installed to `FLUX_SHELL_RC_PATH`.

On a non-spack system, the `mpibind.lua` file can just be installed directly into the default rcpath, e.g. `/etc/flux/shell/lua.d`.

### CODE REVIEW

**Comment by garlick** [comment]
That seems like it would work. 

So shell plugins supplied by external packages would supply a `packagename.lua` script.

The flux pmix  plugin should probably be enabled/disabled by the mpi personalities.  Is there a way for  `openmpi.lua` and `spectrum.lua` to set an option that a `flux-pmix.lua` script could test?

**Comment by grondo** [comment]
> So shell plugins supplied by external packages would supply a packagename.lua script.

If you want your shell plugin always loaded, the `packagename.lua` is optional if you are not in an environment like spack - you can just install the `.so` to the default `fluxshellpluginpath`. Otherwise, yes an external package would install a `package.lua` which could then optionally or always load the plugin.

> The flux pmix plugin should probably be enabled/disabled by the mpi personalities. Is there a way for openmpi.lua and spectrum.lua to set an option that a flux-pmix.lua script could test?

They could, though I'm not sure if it is good practice for flux-core builtin mpi personalities to assume existence of any external packages (this would require an external package developer to get some change into flux-core before their package would work (?))

Can the `flux-pmix.lua` rc script be considered an "extension" of the built-in MPI personalities, and take action independently if `shell.options.mpi` is set to some set of supported MPIs?

**Comment by grondo** [comment]
It occurs to me that one issue here is that this laissez faire approach to deferring the handling of `shell.options.mpi` is that an invalid option is ignored. E.g. I can request the `awesome` MPI personality explicitly, and my plea will be silently ignored:

```
$ flux mini run -o mpi=awesome hostname
asp
```

Eventually, we may want to make some shell options first class and handled specifically in the default initrc.
E.g. If `-o mpi` is used, the shell could try to load a specific plugin from `${rcpath}/mpi/${name}.lua`, then if a version is provided, try `mpi/${name}-version.lua`. If neither of these files are found, it could be an error.

**Comment by grondo** [comment]
Another thought -- since this issue really only affects the spack package of flux-core, another solution would be to modify the default shell initrc similarly in the flux-core spack package. Then no changes here would be necessary.

One benefit of this change though is that you can more easily guarantee order of rc scripts during shell startup, since the FLUX_SHELL_RC_PATH directories are traversed in order (starting with the built in one)

**Comment by garlick** [comment]
> One benefit of this change though is that you can more easily guarantee order of rc scripts during shell startup, since the FLUX_SHELL_RC_PATH directories are traversed in order (starting with the built in one)

Yeah, this change seems good to me as is.

**Comment by grondo** [comment]
@SteVwonder or @eleon, do you concur that this will be a good enough solution for #3866?

If I can get an ACK or review from one of you then I will set MWP.

**Comment by eleon** [comment]
Thank you for the prompt solution, @grondo. This looks good to me. I will try it out with `mpibind` in Spack when I get an opportunity and report back.

**Comment by grondo** [comment]
Thanks @eleon. I will set this PR to merge. Once that is done if you build flux@master in spack you should have this feature. If there are any issues, please open another GitHub issue and we'll address from there. 

Thank you!

**Comment by eleon** [comment]
Will do, @grondo. Thank you!

**Comment by grondo** [comment]
Oh still waiting for an officieal review from @SteVwonder or @garlick

**Review by garlick — APPROVED** [review]
LGTM!

**Comment by grondo** [comment]
Thanks!

**Comment by eleon** [comment]
@grondo , nice work! I successfully installed mpibind+flux through Spack: 
```
spack install mpibind@master+cuda+flux ^lua@5.1.4
```
```
$ flux mini run -n2 -c18 ./affinity/mpi
pascal25   Task   0/  2 runing on 9 CPUs: 0-8
           Task   0/  2 has 1 GPUs: 0x4 
pascal25   Task   1/  2 runing on 9 CPUs: 9-17
           Task   1/  2 has 1 GPUs: 0x7 
```
A few notes: 
* I needed a specific version of `lua` that matches my Slurm configuration
* The updated mpibind recipe is not upstream yet
* `mpibind` works out of the box with Flux (but can be disabled with `-o mpibind=off`)!

Question: 
* Is it safe to assume that the new environment variable is present starting with Flux version 0.29?

**Comment by grondo** [comment]
> Is it safe to assume that the new environment variable is present starting with Flux version 0.29?

No, unfortunately it will only be supported with Flux > 0.29.0

**Comment by eleon** [comment]
Okay. I will leave the spack recipe with `flux@master` for now and change to '>0.29' when the Flux spack recipe is updated.

**Pull request merged**
